### ISSUE TITLE
Extension: Getting config setting after waiting for update doesn't provide updated value

### ISSUE THREAD

**Issue by eamodio** [Microsoft/vscode#25508]
- VSCode Version: Code - Insiders 1.12.0-insider (48ae6e3027be560d45dd1ebfbe090d6ff77af7f0, 2017-04-26T06:13:08.212Z)
- OS Version: Windows_NT ia32 10.0.16179
---

Steps to Reproduce:
1. Create an extension that has a setting and does the following
```ts
// Get the current value of the setting -- lets say its 'foo'
const setting = vscode.workspace.getConfiguration('extension').get('setting');

// Wait for the update to complete (doesn't matter if its async/await or .then)
await vscode.workspace.getConfiguration('extension').update('setting', 'bar');

// Here I would expect updatedSetting to equal 'bar', but instead it will be 'foo'
const updatedSetting = vscode.workspace.getConfiguration('extension').get('setting');
```
2. Observe that updatedSetting does not return the updated value

//cc @jrieken

**Comment by jrieken** [comment]
From the extension host we call [`this._configurationEditingService.writeConfiguration`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/mainThreadConfiguration.ts#L38) and the expectation is that the promise it returns doesn't resolve before a configuration change event is fired. I see some queuing and it seems this behaviour has changed, effectively breaking extensions

**Comment by sandy081** [comment]
I think this is the case always.. Configuration editing service was queuing always (AFIK) and never waited for event to trigger. Its the same behaviour for core and extensions. Everybody listens to onConfigurationChange event for any changes to configurations. Definitely, editing service can be improved to resolve after configuration is updated both on disk and in memory. So not sure if is a candidate for end game.

**Comment by eamodio** [comment]
I could certainly be mistaken, but I thought at one point (quite a while ago), I would get the event before the promise resolved (which would be ideal imo) -- or at least I have some recollection of having to change my code to deal with that (though it could have just been a bug I didn't notice before). But I'm pretty sure that I used to be able to query for the value after waiting for the update to get the most recent value (even without the event firing).

And without this guarantee it makes the code an extension needs to write to wait, very error prone, since if you need to wait for the event it gets challenging -- since you aren't guaranteed to get the event (depending on what you are updating). This cause me to need to write ugly code like: https://github.com/eamodio/vscode-toggle-excluded-files/blob/master/src/excludeController.ts#L154

I believe this is also causing some intermittent timing issues I am seeing in GitLens as well

**Comment by jrieken** [comment]
Yes, it should fire the event and update the state before the promise resolves. We try to figure when that got lost...

**Comment by sandy081** [comment]
Sorry guys.. It was working before and it was broken by me in 1.11. Here is the change that caused this - 4c2600fd3f5f41739ec97951755e94b83165a04c

**Comment by jrieken** [comment]
Still seeing this

**Comment by jrieken** [comment]
Works now

**Comment by sandy081** [comment]
I actually fixed it in configuration editing service to reload configuration after editing. Reload configuration should trigger change event. Global configuration is triggering but workspace configurations is not. Refactorings from very long back (1.4) changed the behaviour of workspace configuration service. It is not triggering change event anymore.

Fixed workspace configuration to trigger event in reload.
